File: app/src/main/java/com/example/workoutapp/di/AppComponent.java
package com.example.workoutapp.di;

import android.app.Application;

import com.example.workoutapp.WorkoutApplication;
import com.example.workoutapp.di.addroutine.AddRoutineComponent;
import com.example.workoutapp.di.addroutine.AddRoutineModule;
import com.example.workoutapp.di.addworkout.AddWorkoutComponent;
import com.example.workoutapp.di.addworkout.AddWorkoutModule;
import com.example.workoutapp.di.showroutine.ShowRoutineComponent;
import com.example.workoutapp.di.showroutine.ShowRoutineModule;
import com.example.workoutapp.di.showworkout.ShowWorkoutComponent;
import com.example.workoutapp.di.showworkout.ShowWorkoutModule;

import dagger.BindsInstance;
import dagger.Component;

/**
 * if you wanna make a new feature:
 * <p>
 * - make a new component interface to inject to your new activities and fragments
 * - make a new module to describe how to create the dependencies in your activities and fragments
 * <p>
 * - add a plus method to this app component
 * - add a create method to FlashCardComponentProvider
 * <p>
 * the rest was groundwork that makes it easier to work with Dagger
 */

@Component(modules = {AndroidModule.class, CompositeDisposableModule.class})
public interface AppComponent {
    void inject(WorkoutApplication workoutApplication);

    AddWorkoutComponent plus(AddWorkoutModule addWorkoutModule);

    ShowWorkoutComponent plus(ShowWorkoutModule showWorkoutModule);

    AddRoutineComponent plus(AddRoutineModule addRoutineModule);

    ShowRoutineComponent plus(ShowRoutineModule showRoutineModule);

    @Component.Builder
    interface Builder {
        @BindsInstance
        Builder application(Application application);

        AppComponent build();
    }
}

File: app/src/main/java/com/example/workoutapp/di/AppComponent.kt
package com.example.workoutapp.di

import com.example.workoutapp.di.addroutine.AddRoutineModule
import com.example.workoutapp.di.addworkout.AddWorkoutModule
import com.example.lib_key_value_storage.di.android.SharedPreferencesModule
import com.example.workoutapp.di.api.ApiModule
import com.example.workoutapp.di.app.AppModule
import com.example.workoutapp.di.home.HomeModule
import com.example.workoutapp.di.login.LoginModule
import com.example.workoutapp.di.profile.ProfileModule
import com.example.workoutapp.di.register.RegisterModule
import com.example.workoutapp.di.showroutine.ShowRoutineModule
import com.example.workoutapp.di.showworkout.ShowWorkoutModule
import com.example.workoutapp.di.splash.SplashModule
import com.example.workoutapp.ui.WorkoutApplication
import dagger.Component
import dagger.android.AndroidInjector
import dagger.android.support.AndroidSupportInjectionModule
import javax.inject.Singleton

@Singleton
@Component(
    modules = [
        AppModule::class,
        AndroidSupportInjectionModule::class,
        com.example.core.di.CompositeDisposableModule::class,
        AddRoutineModule::class,
        AddWorkoutModule::class,
        ApiModule::class,
        HomeModule::class,
        LoginModule::class,
        ProfileModule::class,
        RegisterModule::class,
        SharedPreferencesModule::class,
        ShowRoutineModule::class,
        ShowWorkoutModule::class,
        SplashModule::class,
        ActivityModule::class,
        ViewModelModule::class
    ]
)
interface AppComponent : AndroidInjector<WorkoutApplication> {

    @Component.Factory
    abstract class Factory : AndroidInjector.Factory<WorkoutApplication>

}